internal/indexer/documentation.go

```diff
@@ -0,0 +1,748 @@
+	label string
+
+	// The name of the type.
+	name string
+
+	// The full type signature, with docstrings on e.g. methods and struct fields.
+	signature string
+
+	// Documentation strings in Markdown.
+	docsMarkdown string
+
+	// Is the type itself exported, deprecated?
+	exported, deprecated bool
+
+	// The type itself.
+	typ types.Type
+}
+
+func (t typeDocs) result() *documentationResult {
+	var tags []protocol.DocumentationTag
+	if t.exported {
+		tags = append(tags, protocol.DocumentationExported)
+	}
+	if t.deprecated {
+		tags = append(tags, protocol.DocumentationDeprecated)
+	}
+
+	// Include the full type signature
+	var detail bytes.Buffer
+	fmt.Fprintf(&detail, "```Go\n")
+	fmt.Fprintf(&detail, "%s\n", t.signature)
+	fmt.Fprintf(&detail, "```\n\n")
+	fmt.Fprintf(&detail, "%s", t.docsMarkdown)
+
+	return &documentationResult{
+		Documentation: protocol.Documentation{
+			Slug:    slugify(t.name),
+			NewPage: false,
+			Tags:    tags,
+		},
+		Label:  protocol.NewMarkupContent(t.label, protocol.PlainText),
+		Detail: protocol.NewMarkupContent(detail.String(), protocol.Markdown),
+	}
+}
+
+func (d *docsIndexer) indexTypeSpec(p *packages.Package, in *ast.TypeSpec, isTestFile bool) typeDocs {
+	var result typeDocs
+	result.label = fmt.Sprintf("type %s %s", in.Name.String(), formatTypeLabel(p.TypesInfo.TypeOf(in.Type)))
+	result.name = in.Name.String()
+	result.typ = p.TypesInfo.ObjectOf(in.Name).Type()
+	result.exported = ast.IsExported(in.Name.String()) && !isTestFile
+	result.deprecated = isDeprecated(in.Doc.Text())
+
+	// Produce the full type signature with docs on e.g. methods and struct fields, but not on the
+	// type itself (we'll produce those as Markdown below.)
+	cpy := *in
+	cpy.Doc = nil
+	result.signature = "type " + formatNode(p.Fset, &cpy)
+
+	result.docsMarkdown = godocToMarkdown(in.Doc.Text())
+	return result
+}
+
+type funcDocs struct {
+	// The emitted "documentationResult" vertex ID.
+	ID uint64
+
+	// The best one-line label for this type we could come up with, e.g. `func foo (e struct{...})`
+	// omitting field names.
+	label string
+
+	// The full type signature, with docstrings on e.g. methods and struct fields.
+	signature string
+
+	// The name of the function.
+	name string
+
+	// Documentation strings in Markdown.
+	docsMarkdown string
+
+	// Is the type itself exported, deprecated?
+	exported, deprecated bool
+
+	// The type of the receiver, or nil.
+	recvType ast.Expr
+
+	// The type of return values, or nil.
+	resultTypes []ast.Expr
+}
+
+func (f funcDocs) result() *documentationResult {
+	var tags []protocol.DocumentationTag
+	if f.exported {
+		tags = append(tags, protocol.DocumentationExported)
+	}
+	if f.deprecated {
+		tags = append(tags, protocol.DocumentationDeprecated)
+	}
+
+	// Include the full type signature
+	var detail bytes.Buffer
+	fmt.Fprintf(&detail, "```Go\n")
+	fmt.Fprintf(&detail, "%s\n", f.signature)
+	fmt.Fprintf(&detail, "```\n\n")
+	fmt.Fprintf(&detail, "%s", f.docsMarkdown)
```

Comment:
```suggestion
	var detail strings.Builder
	detail.WriteString("```Go\n")
	detail.WriteString("%s\n", f.signature)
	detail.WriteString("```\n\n")
	detail.WriteString("%s", f.docsMarkdown)
```

using strings.Builder would most likely be more efficient both from memory and cpu perspective here than a bytes.Buffer through fmt.Fprintf; calls to the bytes.Buffer method are probably virtualized as fmt.Fprintf takes an io.Writer interface, and the strings passed might be copied as they are converted to []byte, but I havent confirmed that last part. With strings.Builder, there is some magic to avoid copying of the string.

Even better I believe would be basic string concatenation, as these are straight combining (as opposed to more complex control flow) so the compiler can most likely make this an optimal operation. I will write some benchmarks out of curiosity :slightly_smiling_face:

Reply:
Without `b.Grow(...)`, string concat outperforms bytes.Buffer and strings.Builder, but with it, strings.Builder comes out on top 
![image](https://user-images.githubusercontent.com/18282288/117984528-d39c4200-b32f-11eb-91fa-4617be10925a.png)

Reply:
```suggestion
	var detail strings.Builder
	detail.Grow(6 + len(f.signature) + len(f.docsMarkdown) + 6)
	detail.WriteString("```Go\n")
	detail.WriteString(f.signature)
	detail.WriteRune('\n')
	detail.WriteString("```\n\n")
	detail.WriteString(f.docsMarkdown)
```

Reply:
Thank you!